### PR TITLE
Fix EM_ASM for wasm backend

### PR DESCRIPTION
**Pull request by jgravelle-google**
Avoids using varargs for EM_ASM implementation functions so that later compiler magic (s2wasm) can extract correct signature metadata.

Doesn't wind up changing how this is done for asmjs and asm2wasm.
Depends on https://github.com/WebAssembly/binaryen/pull/910 for s2wasm support.

### CODE REVIEW

File: system/include/emscripten/em_asm.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#ifndef __em_asm_h__
+#define __em_asm_h__
+
+#ifndef __cplusplus
+void emscripten_asm_const();
```
Comment (dschuff) [inline system/include/emscripten/em_asm.h]: Should probably have a comment that explicitly calls out that these are non-prototype declarations whose purpose is to be valid ISO C (which requires declarations) while still allowing non-vararg calls with different parameters.

File: system/include/emscripten/em_asm.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+double emscripten_asm_const_double();
+#else
+#ifdef __asmjs
+extern "C" {
+    void emscripten_asm_const(const char* code);
+    int emscripten_asm_const_int(const char* code, ...);
+    double emscripten_asm_const_double(const char* code, ...);
+}
+#else
+template <typename... Args> void emscripten_asm_const(const char* code, Args...);
```
Comment (dschuff) [inline system/include/emscripten/em_asm.h]: and maybe a counterpart comment here too

File: system/include/emscripten/em_asm.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#ifndef __em_asm_h__
```
Comment (dschuff) [inline system/include/emscripten/em_asm.h]: need copyright header?

Comment (jgravelle-google) [inline system/include/emscripten/em_asm.h]: Actually none of the other headers in the `system/include/emscripten` folder seem to have them. @kripken is that a thing?

Comment (kripken) [inline system/include/emscripten/em_asm.h]: you're right, we just have the toplevel copyright, so i think this is fine

**Comment by kripken** [comment]
Neat ;)

This looks like it should work to me. I'd verify that `ALL.test_inline*`, `test*em_asm*` (note two stars) and all of `other` pass.

**Comment by jgravelle-google** [comment]
All the relevant tests are passing and binaryen-side this is good; am I good to merge this?

**Comment by kripken** [comment]
Sure, but please keep an eye on bots for unexpected issues.

**Comment by jgravelle-google** [comment]
Good point; I'll commit tomorrow morning then so I can keep an eye on them longer

**Pull request merged**